### PR TITLE
executor: kill tableReader for each connection correctly (#18277)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #18277 to release-3.0

---

### What problem does this PR solve?

Issue Number: close a part of #18057 

Problem Summary:
The tableReader in the coprocessor may retry many times to read regions, druing which, it cannot be aware of the cancelation from client. It may take much of memory for a long time. 

### What is changed and how it works?

Proposal: detect `Killed` to stop the goroutine.

What's Changed: should not use `CompareAndSwap` to change the `Killed` before Close(), because many runninig goroutines should detect the `Killed`.

How it Works: the TableReader the detect the `killed` in every loop.
The query may not exit quickly, because the query do some close work.

### Check List 

Tests 

- Integration test 
- Manual test (add detailed scripts or steps below)
test on mocktikv:
```
step1. explain analyze select * from lineitem;
step2. kill tidb sessionID on another client;
to see the connection is cancelled.
```
similarly test on TiFlash is ok.

### Release note 

- To speed up canceling a query

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -385,6 +385,9 @@ func (s *Server) onConn(conn *clientConn) {
 	if plugin.IsEnable(plugin.Audit) {
 		conn.ctx.GetSessionVars().ConnectionInfo = conn.connectInfo()
 	}
+	type vars struct {
+		killed *uint32
+	}
```
Comment (qw4990) [inline server/server.go:390]: It seems these lines can be removed.

Comment (fzhedu) [reply server/server.go]: > It seems these lines can be removed.

Done

**Pull request merged**